src/integrations/stripe.ts

```diff
@@ -138,6 +195,92 @@ export async function setupPaymentMethod(
   return null;
 }
 
+export function isProductSelected({
+  productId,
+  selectedProductIdField,
+  fieldValues
+}: {
+  productId: string;
+  selectedProductIdField: string;
+  fieldValues: { [_: string]: any };
+}) {
+  if (productId && selectedProductIdField) {
+    // get the value from the hidden field.  Example: {"some_stripe_product_id": 1}
+    // selectedProductIdField can be shared, so make sure it is our product that is selected.
+    return Boolean((fieldValues[selectedProductIdField] || {})[productId]);
+  }
+  return false;
+}
+
+export async function toggleProductSelection({
+  productId,
+  selectedProductIdFieldId,
+  selectedProductIdFieldKey,
+  fieldValues,
+  updateFieldValues,
+  client,
+  integrations
+}: {
+  productId: string;
+  selectedProductIdFieldId: string;
+  selectedProductIdFieldKey: string;
+  fieldValues: { [_: string]: any };
+  updateFieldValues: any;
+  client: any;
+  integrations: any;
+}) {
+  // check productId and selectedProductIdField (key) are set
+  if (
+    productId &&
+    selectedProductIdFieldKey &&
+    selectedProductIdFieldId &&
+    integrations.stripe
+  ) {
+    // get the value from the hidden field
+    const fieldVal = fieldValues[selectedProductIdFieldKey] || {};
+    const quantity = fieldVal[productId];
+    let newQuantity = quantity ^ 1;
+    // For the single select case, the stripe product id will change in the
+    // common/shared hidden field. In this case just replace with quantity 1
+    // of the new product id.
+    if (!(productId in fieldVal)) newQuantity = 1;
+
+    // toggle the quantity and set the new value into the hidden field
+    updateFieldValues({
+      [selectedProductIdFieldKey]: {
+        [productId]: newQuantity
+      }
+    });
+
+    // set the new product selection state on the BE
+    const { total_cost: total, field_update: fieldUpdate } =
+      await client.updateProductSelection(
+        productId,
+        newQuantity,
+        selectedProductIdFieldId
+      );
+
+    // set the RETURNED value into the hidden field so that BE is source of truth
+    updateFieldValues({
+      [fieldUpdate.field_key]: {
+        [fieldUpdate.value.stripe_product_id]: fieldUpdate.value.quantity
+      }
+    });
+
+    // set the total into the total hidden field if mapped (leaving the currency symbol off)
+    const totalFieldKey =
```

Comment:
this logic seems like it could be part of the client's return value for updateProductSelection. ideally couldn't we have updateProductSelection returna. dict of field values to update directly?